Prática I

Eduardo Garcia Misiuk - 9293230 
Mauricio Caetano da Silva - 9040996


Datasets escolhidos: Titanic e Instituições de Ensino Básico

Primeira parte: Titanic

1) Hipóteses
        1) A classe de passagem influenciou na sobrevivência das pessoas?
        2) O gênero das pessoas influenciou na chance delas sobreviverem?
        3) A faixa de idade das pessoas foi um fator determinante para sua sobrevivência? (independentemente de gênero)      


2) Atributos

            A) Tipo de passagem (Pclass)
                Tipo de dado: Numérico racional discreto
                Operações: Média, moda, mediana.
                
            B) Gênero (Gender)
                Tipo de dado: Simbólico nominal
                Operações: Moda.
                
            C) Idade (Age)
                Tipo de dado: Numérico racional contínuo
                Operações: Média, moda, mediana.
                
            D) Parentes a bordo (SibSp)
                Tipo de dado: Numérico racional discreto
                Operações: Média, moda, mediana.  
                
            E) Tarifa (Fare)
                Tipo de dado: Numérico racional contínuo
                Operações: Média, moda, mediana.
                                
                
       Os atributos dos itens A e E (Pclass e Fare, respectivamente) acima estão relacionados à pergunta 1 do item anterior, pois elas descrevem qual o tipo de passagem e qual o valor pago por um determinado passageiro. De modo similar, a hipótese 2 está relacionada diretamente com o atributo B (Gender), assim como a hipótese 3 está relacionada ao atributo C (Age) por motivos óbvios.
       O atributo D, irmãos/cônjuges abordo, apesar de não se relacionar diretamente com as hipóteses anteriores, foi escolhido também pois suspeitamos que esse fato pode ter influenciado nas mortes do acidente.

In [39]:
# Questão 3, análise exploratória do dataset Titanic
# Abrindo o arquivo e selecionando as colunas de interesse
import numpy as np
import pandas as pd
from scipy import stats

# Colunas de interesse - atributos
att = ['PassengerId','Survived','Pclass','Sex','Age','SibSp','Fare']

data = pd.read_csv('titanic.csv', usecols=att)
data


PassengerId  Survived  Pclass     Sex   Age  SibSp      Fare
0              1         0       3    male  22.0      1    7.2500
1              2         1       1  female  38.0      1   71.2833
2              3         1       3  female  26.0      0    7.9250
3              4         1       1  female  35.0      1   53.1000
4              5         0       3    male  35.0      0    8.0500
5              6         0       3    male   NaN      0    8.4583
6              7         0       1    male  54.0      0   51.8625
7              8         0       3    male   2.0      3   21.0750
8              9         1       3  female  27.0      0   11.1333
9             10         1       2  female  14.0      1   30.0708
10            11         1       3  female   4.0      1   16.7000
11            12         1       1  female  58.0      0   26.5500
12            13         0       3    male  20.0      0    8.0500
13            14         0       3    male  39.0      1   31.2750
14            15         0       3  female  14.0      0    7.8542
15            16         1       2  female  55.0      0   16.0000
16            17         0       3    male   2.0      4   29.1250
17            18         1       2    male   NaN      0   13.0000
18            19         0       3  female  31.0      1   18.0000
19            20         1       3  female   NaN      0    7.2250
20            21         0       2    male  35.0      0   26.0000
21            22         1       2    male  34.0      0   13.0000
22            23         1       3  female  15.0      0    8.0292
23            24         1       1    male  28.0      0   35.5000
24            25         0       3  female   8.0      3   21.0750
25            26         1       3  female  38.0      1   31.3875
26            27         0       3    male   NaN      0    7.2250
27            28         0       1    male  19.0      3  263.0000
28            29         1       3  female   NaN      0    7.8792
29            30         0       3    male   NaN      0    7.8958
..           ...       ...     ...     ...   ...    ...       ...
861          862         0       2    male  21.0      1   11.5000
862          863         1       1  female  48.0      0   25.9292
863          864         0       3  female   NaN      8   69.5500
864          865         0       2    male  24.0      0   13.0000
865          866         1       2  female  42.0      0   13.0000
866          867         1       2  female  27.0      1   13.8583
867          868         0       1    male  31.0      0   50.4958
868          869         0       3    male   NaN      0    9.5000
869          870         1       3    male   4.0      1   11.1333
870          871         0       3    male  26.0      0    7.8958
871          872         1       1  female  47.0      1   52.5542
872          873         0       1    male  33.0      0    5.0000
873          874         0       3    male  47.0      0    9.0000
874          875         1       2  female  28.0      1   24.0000
875          876         1       3  female  15.0      0    7.2250
876          877         0       3    male  20.0      0    9.8458
877          878         0       3    male  19.0      0    7.8958
878          879         0       3    male   NaN      0    7.8958
879          880         1       1  female  56.0      0   83.1583
880          881         1       2  female  25.0      0   26.0000
881          882         0       3    male  33.0      0    7.8958
882          883         0       3  female  22.0      0   10.5167
883          884         0       2    male  28.0      0   10.5000
884          885         0       3    male  25.0      0    7.0500
885          886         0       3  female  39.0      0   29.1250
886          887         0       2    male  27.0      0   13.0000
887          888         1       1  female  19.0      0   30.0000
888          889         0       3  female   NaN      1   23.4500
889          890         1       1    male  2

In [40]:
# Análise para Tipo de Passagem
Pclass = data['Pclass']

# Média
avrg = np.average(Pclass)
print("Media: ", avrg)

# Média podada
trimmed = stats.trim_mean(Pclass, proportiontocut=0.20)
print("Trimmed: ", trimmed)

# Winsorização
wins = np.average(stats.mstats.winsorize(Pclass, limits=0.20))
print("Wins: ", wins)

# Mediana
median = np.median(Pclass)
print("Mediana: ", median)

# Moda
mode = stats.mode(Pclass)
print("Moda: ", mode.mode)

Media:  2.308641975308642
Trimmed:  2.514018691588785
Wins:  2.308641975308642
Mediana:  3.0
Moda:  [3]


    Média, moda, mediana e as médias podadas ajudam a entender qual a condição financeira da maioria dos passageiros do navio, uma vez que o atributo Pclass determina a classe de passagem comprada. Pclass assume os valores 1, 2 e 3 para primeira, segunda e terceira classe, respectivamente.
    Dessa forma, pode-se perceber que existiam mais passageiros de segunda e terceira classes em relação aos passageiros de primeira classe.

In [41]:
# Análise para gênero
sex = data['Sex']

# Moda
mode = stats.mode(sex)
print("Moda: ", mode.mode)

Moda:  ['male']


/home/misiuk/anaconda3/lib/python3.6/site-packages/scipy/stats/stats.py:245: RuntimeWarning: The input array could not be properly checked for nan values. nan values will be ignored.
  "values. nan values will be ignored.", RuntimeWarning)


    Para o atributo de gênero (sex), a medida que faz mais sentido é a moda. Através da análise desse atributo pode ser visto que existiam mais homens do que mulheres no navio.

In [42]:
# Análise para Idade
age = data['Age'][~np.isnan(data['Age'])] # Seleciona apenas os valores de idade diferentes de NaN

# Média
avrg = np.average(age)
print("Media: ", avrg)

# Média podada
trimmed = stats.trim_mean(age, proportiontocut=0.20)
print("Trimmed: ", trimmed)

# Winsorização
wins = np.average(stats.mstats.winsorize(age, limits=0.20))
print("Wins: ", wins)

# Mediana
median = np.median(age)
print("Mediana: ", median)

# Moda
mode = stats.mode(age)
print("Moda: ", mode.mode)

Media:  29.69911764705882
Trimmed:  28.663953488372094
Wins:  29.195378151260503
Mediana:  28.0
Moda:  [24.]


    Nós achamaos que a idade também pode ser um fator determinante de sobrevivência dos passageiros, uma vez que pode ter sido determinada preferência de salvamento das pessoas conforme sua idade. Através da observação dos valores acima podemos dizer que as pessoas do navio tinham aproximadamente 30 anos.

In [43]:
# Análise para número de irmãos / cônjuges
sib = data['SibSp'];

# Média
avrg = np.average(sib)
print("Media: ", avrg)

# Média podada
trimmed = stats.trim_mean(sib, proportiontocut=0.20)
print("Trimmed: ", trimmed)

# Winsorização
wins = np.average(stats.mstats.winsorize(sib, limits=0.20))
print("Wins: ", wins)

# Mediana
median = np.median(sib)
print("Mediana: ", median)

# Moda
mode = stats.mode(sib)
print("Moda: ", mode.mode)

Media:  0.5230078563411896
Trimmed:  0.19626168224299065
Wins:  0.3176206509539843
Mediana:  0.0
Moda:  [0]


    Nesse caso é possível observar que a maioria das pessoas não possuiam irmãos / cônjuges a bordo. Esse fator também pode ser importante para determinar se uma pessoa sobrevive ou não.

In [44]:
# Análise para Tarifa
fare = data['Fare'];

# Média
avrg = np.average(fare)
print("Media: ", avrg)

# Média podada
trimmed = stats.trim_mean(fare, proportiontocut=0.20)
print("Trimmed: ", trimmed)

# Winsorização
wins = np.average(stats.mstats.winsorize(fare, limits=0.20))
print("Wins: ", wins)

# Mediana
median = np.median(fare)
print("Mediana: ", median)

# Moda
mode = stats.mode(fare)
print("Moda: ", mode.mode)

Media:  32.204207968574636
Trimmed:  17.45484953271028
Wins:  19.97841425364759
Mediana:  14.4542
Moda:  [8.05]


    Assim como no caso da classe da passagem, nós supomos que as tarifas que o passageiro pagou poderiam ter afetado no seu tratamento quando ocorreu a emergência. 
    Nesse caso, observamos uma certa diferença entre os valores da média comum para as médias podadas. Isso ocorre devido a presença de valores discrepantes ("outliers") de tarifas pagas por alguns poucos passageiros.

4) Com relação às hipóteses levantadas, apesar das operações retornarem dados concretos e possibilitarem algumas conclusões superficiais sobre o acidente, não podemos responder às perguntas levantadas no início do item 1, uma vez que os dados foram analisados separadamente, não havendo correspondências com o fato de uma determinada pessoa sobreviver ou não.

5) Podemos destacar como um problema o fato de alguns valores de idade não existirem, desse modo, algumas células da tabela de dados possuiam 'NaN' ("Not a number") como valor de idade. Desse modo, foi realizada uma filtragem desses dados, como visto na linha 4 do quadro referente à idade dos passageiros, eliminando as instâncias com células inválidas.

Parte 2: Instituições de Ensino

1) Hipóteses
        1) A grande maioria das escolas tem água filtrada.
        2) Qual a porcentagem de escolas que não têm energia?
        3) Qual a porcentagem de escolas que oferecem alimentação?

2) Atributos

            A) Inexistência de água (AGUA_INEXISTENTE)
                Tipo de dado: nominal
                Operações: moda, porcentagem, igual (==), diferente (!=)

            B) Existência de água filtrada (AGUA_FILTRADA)
                Tipo de dado: nominal
                Operações: moda, porcentagem, igual (==), diferente (!=)

            C) Inexistência de energia elétrica (ENERGIA_INEXISTENTE)
                Tipo de dado: nominal
                Operações: moda, porcentagem, igual (==), diferente (!=)

            D) Inexistência de esgoto (ESGOTO_INEXISTENTE)
                Tipo de dado: nominal
                Operações: moda, porcentagem, igual (==), diferente (!=)

            E) Existência de alimentação (ALIMENTACAO)
                Tipo de dado: nominal
                Operações: moda, porcentagem, igual (==), diferente (!=)

       Os atributos acima foram escolhidos pois conseguimos os dados necessários para responder a perguntas sobre o oferecimento de serviços básicos às escolas de São Paulo.

In [45]:
# Questão 3, análise exploratória do dataset Instituições de Ensino
# Abrindo o arquivo e selecionando as colunas de interesse
import numpy as np
import pandas as pd
from scipy import stats

att = ['ALIMENTACAO', 'AGUA_FILTRADA', 'AGUA_INEXISTENTE', 'ENERGIA_INEXISTENTE', 'ESGOTO_INEXISTENTE']

# O arquivo está em unicode e os separadores são ';', portanto é preciso mandar essa informação à função
data = pd.read_csv('escolas.csv', usecols=att, sep=';', encoding='ISO-8859-1')

# Removendo linhas inválidas
data = data.dropna()

data

ALIMENTACAO AGUA_FILTRADA AGUA_INEXISTENTE ENERGIA_INEXISTENTE  \
0             Sim      Filtrada              Não                 Não   
1             Sim      Filtrada              Não                 Não   
2             Sim      Filtrada              Não                 Não   
3             Sim      Filtrada              Não                 Não   
4             Sim      Filtrada              Não                 Não   
5             Sim      Filtrada              Não                 Não   
6             Sim      Filtrada              Não                 Não   
7             Sim      Filtrada              Não                 Não   
8             Sim      Filtrada              Não                 Não   
9             Sim      Filtrada              Não                 Não   
10            Sim      Filtrada              Não                 Não   
11            Sim      Filtrada              Não                 Não   
12            Sim      Filtrada              Não                 Não   
13            Sim      Filtrada              Não                 Não   
14            Sim      Filtrada              Não                 Não   
15            Sim      Filtrada              Não                 Não   
16            Sim      Filtrada              Não                 Não   
17            Sim      Filtrada              Não                 Não   
18            Sim      Filtrada              Não                 Não   
19            Sim      Filtrada              Não                 Não   
20            Sim      Filtrada              Não                 Não   
21            Sim      Filtrada              Não                 Não   
22            Sim      Filtrada              Não                 Não   
23            Sim      Filtrada              Não                 Não   
24            Sim      Filtrada              Não                 Não   
25            Sim      Filtrada              Não                 Não   
26            Sim      Filtrada              Não                 Não   
27            Sim      Filtrada              Não                 Não   
28            Sim      Filtrada              Não                 Não   
29            Sim      Filtrada              Não                 Não   
...           ...           ...              ...                 ...   
31029         Sim      Filtrada              Não                 Não   
31030         Sim      Filtrada              Não                 Não   
31031         Sim      Filtrada              Não                 Não   
31032         Sim      Filtrada              Não                 Não   
31033         Sim      Filtrada              Não                 Não   
31034         Sim      Filtrada              Não                 Não   
31035         Sim      Filtrada              Não                 Não   
31036         Sim      Filtrada              Não                 Não   
31037         Sim      Filtrada              Não                 Não   
31038         Sim      Filtrada              Não                 Não   
31039         Sim      Filtrada              Não                 Não   
31040         Sim      Filtrada              Não                 Não   
31041         Sim      Filtrada              Não                 Não   
31042         Sim      Filtrada              Não                 Não   
31043         Sim      Filtrada              Não                 Não   
31044         Sim      Filtrada              Não                 Não   
31045         Sim      Filtrada              Não                 Não   
31046         Sim      Filtrada              Não                 Não   
31047         Sim      Filtrada              Não                 Não   
31048         Sim      Filtrada              Não                 Não   
31049         Sim      Filtrada              Não                 Não   
31050         Sim      Filtrada              Não                 Não   
31051         Sim      Filtrada              Não                 Não   
31052         Sim      Filtrada              N

In [46]:
# Análise para Alimentação
alimentacao = data['ALIMENTACAO']
n_sim = 0
n_nao = 0

# Moda
mode = stats.mode(alimentacao)
print("Moda: ", mode.mode)

# Análise do número de Sim e Não e suas porcentagens
for x in alimentacao:
    if x == 'Sim':
        n_sim = n_sim + 1
    else:
        n_nao = n_nao + 1

print("Número de Sim: ", n_sim)
print("Número de Não: ", n_nao)
print("Porcentagem de Sim: ", n_sim/(n_sim + n_nao))
print("Porcentagem de Não: ", n_nao/(n_sim + n_nao))

Moda:  ['Sim']
Número de Sim:  22599
Número de Não:  5664
Porcentagem de Sim:  0.7995966457913173
Porcentagem de Não:  0.20040335420868274


/home/misiuk/anaconda3/lib/python3.6/site-packages/scipy/stats/stats.py:245: RuntimeWarning: The input array could not be properly checked for nan values. nan values will be ignored.
  "values. nan values will be ignored.", RuntimeWarning)


In [47]:
# Análise para Água Filtrada
agua_filtrada = data['AGUA_FILTRADA']
n_filtrada = 0
n_nao_filtrada = 0

# Moda
mode = stats.mode(agua_filtrada)
print("Moda: ", mode.mode)

# Análise do número de Sim e Não e suas porcentagens
for x in agua_filtrada:
    if x == 'Filtrada':
        n_filtrada = n_filtrada + 1
    elif x == 'Não filtrada':
        n_nao_filtrada = n_nao_filtrada + 1

print("Número de Filtrada: ", n_filtrada)
print("Número de Não filtrada: ", n_nao_filtrada)
print("Porcentagem de Filtrada: ", n_filtrada/(n_filtrada + n_nao_filtrada))
print("Porcentagem de Não filtrada: ", n_nao_filtrada/(n_filtrada + n_nao_filtrada))

Moda:  ['Filtrada']
Número de Filtrada:  27930
Número de Não filtrada:  333
Porcentagem de Filtrada:  0.9882178112726887
Porcentagem de Não filtrada:  0.011782188727311326


/home/misiuk/anaconda3/lib/python3.6/site-packages/scipy/stats/stats.py:245: RuntimeWarning: The input array could not be properly checked for nan values. nan values will be ignored.
  "values. nan values will be ignored.", RuntimeWarning)


In [48]:
# Análise para Água
agua = data['AGUA_INEXISTENTE']
n_sim = 0
n_nao = 0

# Moda
mode = stats.mode(agua)
print("Moda: ", mode.mode)

# Análise do número de Sim e Não e suas porcentagens
for x in agua:
    if x == 'Sim':
        n_sim = n_sim + 1
    else:
        n_nao = n_nao + 1

print("Número de Sim: ", n_sim)
print("Número de Não: ", n_nao)
print("Porcentagem de Sim: ", n_sim/(n_sim + n_nao))
print("Porcentagem de Não: ", n_nao/(n_sim + n_nao))

Moda:  ['Não']
Número de Sim:  2
Número de Não:  28261
Porcentagem de Sim:  7.076389626012808e-05
Porcentagem de Não:  0.9999292361037398


/home/misiuk/anaconda3/lib/python3.6/site-packages/scipy/stats/stats.py:245: RuntimeWarning: The input array could not be properly checked for nan values. nan values will be ignored.
  "values. nan values will be ignored.", RuntimeWarning)


In [ ]:
# Análise para Energia
energia = data['ENERGIA_INEXISTENTE']
n_sim = 0
n_nao = 0

# Moda
mode = stats.mode(energia)
print("Moda: ", mode.mode)

# Análise do número de Sim e Não e suas porcentagens
for x in energia:
    if x == 'Sim':
        n_sim = n_sim + 1
    else:
        n_nao = n_nao + 1

print("Número de Sim: ", n_sim)
print("Número de Não: ", n_nao)
print("Porcentagem de Sim: ", n_sim/(n_sim + n_nao))
print("Porcentagem de Não: ", n_nao/(n_sim + n_nao))

In [49]:
# Análise para Esgoto
esgoto = data['ESGOTO_INEXISTENTE']
n_sim = 0
n_nao = 0

# Moda
mode = stats.mode(esgoto)
print("Moda: ", mode.mode)

# Análise do número de Sim e Não e suas porcentagens
for x in esgoto:
    if x == 'Sim':
        n_sim = n_sim + 1
    else:
        n_nao = n_nao + 1

print("Número de Sim: ", n_sim)
print("Número de Não: ", n_nao)
print("Porcentagem de Sim: ", n_sim/(n_sim + n_nao))
print("Porcentagem de Não: ", n_nao/(n_sim + n_nao))

Moda:  ['Não']
Número de Sim:  3
Número de Não:  28260
Porcentagem de Sim:  0.00010614584439019212
Porcentagem de Não:  0.9998938541556098


/home/misiuk/anaconda3/lib/python3.6/site-packages/scipy/stats/stats.py:245: RuntimeWarning: The input array could not be properly checked for nan values. nan values will be ignored.
  "values. nan values will be ignored.", RuntimeWarning)


Por serem valores não numéricos, e sim valores textuais, as medidas de moda e mediana não fazem sentido e, por esse motivo, o uso das porcentagens foi necessário. Com as porcentagens conseguimos ter uma visão melhor sobre os valores das variáveis.

4) As questões ainda fazem sentido e, com a análise exploratória, é possível respondê-las. Abaixo estão as respostas.
    1) Esta hipótese foi provada coerente, já que 98.8% das escolas têm água filtrada.
    2) 0.07% das escolas não têm energia.
    3) 79.9% das escolas oferecem alimentação.

5) O problema que tivemos (quanto aos dados) foi a de dados faltantes. Isso foi facilmente contornado retirando tais linhas. Além disso, a planilha contém alguns erros de digitação, como letras faltantes em nomes de atributos.